tools/quic/quic_transport_server.py

```diff
@@ -0,0 +1,244 @@
+        with open(handlers_path + '/' + handler_name) as f:
+            exec(f.read(), global_dict)
+        return EventHandler(self, global_dict)
+
+    def is_closing_or_closed(self) -> bool:
+        if self._quic._close_pending:
+            return True
+        if self._quic._state in END_STATES:
+            return True
+        return False
+
+
+class SessionTicketStore:
+    '''
+    Simple in-memory store for session tickets.
+    '''
+
+    def __init__(self) -> None:
+        self.tickets: Dict[bytes, SessionTicket] = {}
+
+    def add(self, ticket: SessionTicket) -> None:
+        self.tickets[ticket.ticket] = ticket
+
+    def pop(self, label: bytes) -> Optional[SessionTicket]:
+        return self.tickets.pop(label, None)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='QUIC server')
+    parser.add_argument(
+        '-c',
+        '--certificate',
+        type=str,
+        required=True,
+        help='load the TLS certificate from the specified file',
+    )
+    parser.add_argument(
+        '--host',
+        type=str,
+        default='::',
+        help='listen on the specified address (defaults to ::)',
+    )
+    parser.add_argument(
+        '--port',
+        type=int,
+        default=4433,
+        help='listen on the specified port (defaults to 4433)',
+    )
+    parser.add_argument(
+        '-k',
+        '--private-key',
+        type=str,
+        required=True,
+        help='load the TLS private key from the specified file',
+    )
+    parser.add_argument(
+        '--handlers-path',
+        type=str,
+        required=True,
+        help='the directory path of QuicTransport event handlers',
+    )
+    parser.add_argument(
+        '-v',
+        '--verbose',
+        action='store_true',
+        help='increase logging verbosity'
+    )
+    args = parser.parse_args()
+
+    logging.basicConfig(
+        format='%(asctime)s %(levelname)s %(name)s %(message)s',
+        level=logging.DEBUG if args.verbose else logging.INFO,
+    )
+
+    configuration = QuicConfiguration(
+        alpn_protocols=['wq-vvv-01'] + ['siduck'],
+        is_client=False,
+        max_datagram_frame_size=65536,
+    )
+
+    handlers_path = os.path.abspath(os.path.expanduser(args.handlers_path))
+    logging.log(logging.INFO, 'port = %s' % args.port)
+    logging.log(logging.INFO, 'handlers path = %s' % handlers_path)
+
+    # load SSL certificate and key
+    configuration.load_cert_chain(args.certificate, args.private_key)
+
+    ticket_store = SessionTicketStore()
+
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(
+        serve(
+            args.host,
+            args.port,
+            configuration=configuration,
+            create_protocol=QuicTransportProtocol,
+            session_ticket_fetcher=ticket_store.pop,
+            session_ticket_handler=ticket_store.add,
+        )
+    )
+    try:
+        loop.run_forever()
+    except KeyboardInterrupt:
```

Comment:
Question: is there any cleanup that's required? Like releasing the port?

If so, we'd also need to catch `SIGTERM` and `SIGINT` (which is different from `KeyboardInterrupt`, which I think is only raised after interactive Ctrl-C), and exit gracefully.

Reply:
@yutakahirano I took a look and I don't think any cleanup is needed here. There's no subprocess, and the UDP port gets released almost immediately if the process is killed (unlike TCP ports).